src/index.ts

```diff
@@ -106,123 +75,88 @@ function initConfig(projectConfig: Forceable<Config>):
-  } else if (traceAgent.isActive()) {
-    // For unit tests only.
-    // Undoes initialization that occurred last time start() was called.
-    stop();
+export function start(config?: Config): PluginTypes.TraceAgent {
+  const normalizedConfig = initConfig(config || {});
+  // Determine the preferred context propagation mechanism, as
+  // continuation-local-storage should be loaded before any modules that do I/O.
+  if (normalizedConfig.enabled &&
+      normalizedConfig.clsMechanism === 'async-listener') {
+    // This is the earliest we can load continuation-local-storage.
+    require('continuation-local-storage');
   }
 
-  if (!config.enabled) {
-    return traceAgent;
-  }
-
-  const logger = common.logger({
-    level: common.logger.LEVELS[config.logLevel],
-    tag: '@google-cloud/trace-agent'
-  });
-
-  if (modulesLoadedBeforeTrace.length > 0) {
-    logger.error(
-        'TraceAgent#start: Tracing might not work as the following modules',
-        'were loaded before the trace agent was initialized:',
-        `[${modulesLoadedBeforeTrace.sort().join(', ')}]`);
-    // Stop storing these entries in memory
-    filesLoadedBeforeTrace.length = 0;
-    modulesLoadedBeforeTrace.length = 0;
+  if (!tracingSingleton) {
+    tracingSingleton = require('./tracing').tracing;
   }
 
   try {
-    // Initialize context propagation mechanism.
-    const m = config.clsMechanism;
-    const clsConfig: Forceable<TraceCLSConfig> = {
-      mechanism: m === 'auto' ? (useAH ? TraceCLSMechanism.ASYNC_HOOKS :
-                                         TraceCLSMechanism.ASYNC_LISTENER) :
-                                m as TraceCLSMechanism,
-      [FORCE_NEW]: config[FORCE_NEW]
-    };
-    cls.create(clsConfig, logger).enable();
-
-    traceWriter.create(config, logger).initialize((err) => {
-      if (err) {
-        stop();
-      }
-    });
-
-    traceAgent.enable(config, logger);
-
-    pluginLoader.create(config, logger).activate();
+    let tracing: Tracing;
+    try {
+      tracing = tracingSingleton.create(normalizedConfig, {});
+    } catch (e) {
+      // An error could be thrown if create() is called multiple times.
+      // It's not a helpful error message for the end user, so make it more
+      // useful here.
+      throw new Error('Cannot call start on an already created agent.');
+    }
+    tracing.enable();
+    tracing.logModulesLoadedBeforeTrace(filesLoadedBeforeTrace);
+    return tracingSingleton.get().traceAgent;
   } catch (e) {
-    logger.error(
-        'TraceAgent#start: Disabling the Trace Agent for the',
-        `following reason: ${e.message}`);
-    stop();
-    return traceAgent;
-  }
-
-  if (typeof config.projectId !== 'string' &&
-      typeof config.projectId !== 'undefined') {
-    logger.error(
-        'TraceAgent#start: config.projectId, if provided, must be a string.',
-        'Disabling trace agent.');
-    stop();
-    return traceAgent;
+    throw e;
```

Comment:
Just remove the catch block?

Reply:
Done. I think I was about the change the error message but am not sure to what